### ISSUE TITLE
Support ESLint v9

### ISSUE THREAD

**Issue by MikeMcC399** [bmish/eslint-doc-generator#526]
# Request

This is a request for the `eslint-doc-generator` to support usage in an ESLint `v9` environment.

Installing should produce no warnings.

# Why it's needed

ESLint `v9` is the current version. `eslint-doc-generator` assists in generating standardized documentation sections independently of plugin authors. This is possible on ESLint `v8` and earlier versions.

It would be beneficial to users of plugins working under ESLint `v9` to have support for this version. This would enable a continued high standard of plugin documentation information available.

Edit: Since Oct 5, 2024 ESLint `v9` is the lowest supported version of ESLint - see https://eslint.org/version-support/.

## Related issues

- https://github.com/bmish/eslint-doc-generator/issues/376
- https://github.com/bmish/eslint-doc-generator/issues/525
- https://github.com/typescript-eslint/typescript-eslint/issues/8211
- https://github.com/eslint/generator-eslint/pull/179

**Comment by MikeMcC399** [comment]
In the meantime it may be a good idea to add a note to the [README](https://github.com/bmish/eslint-doc-generator/blob/main/README.md) concerning the status with ESLint `v9`.

Currently there is no restriction for ESLint `v9` in the repo according to `package.json`:

https://github.com/bmish/eslint-doc-generator/blob/01c317934c386f699c5eb5f912c6c6b0436201b9/package.json#L86-L88

**Comment by bmish** [comment]
I would like to make sure we work with ESLint v9. However, I'm not actually sure of any ESLint v9 specific issues.

The peer dependency issue in #525 is not really specific to ESLint v9.

Besides that, is there any actual problem using it with ESLint v9?

**Comment by ljharb** [comment]
since eslint 9 defaults to flat config, i assume #376 is a blocker

**Comment by bmish** [comment]
I haven't really identified a specific blocker with flat configs. That issue was more intended to investigate whether flat configs will need additional work to support. As long as flat configs are exported from plugins using the existing `configs` object, we should be able to identify them. The problem is, many plugins are exporting flat configs only via string exports (`const recommended = require('eslint-plugin-foo/configs/recommended');`), and that isn't detectable really. Discussed more in https://github.com/eslint/eslint/issues/18095.

**Comment by MikeMcC399** [comment]
@bmish 

Hello again 😄 

Same confirmation steps as before, running under Windows 11

```text
git clone https://github.com/cypress-io/eslint-plugin-cypress
cd eslint-plugin-cypress
npm ci
npm install eslint-doc-generator@1.7.1 -D
npx eslint-doc-generator
```

No errors apart from the peer dependency warning with `npm install`.

**Comment by MikeMcC399** [comment]
@bmish 

Are you going to pick this up some time? I see that https://typescript-eslint.io/blog/announcing-typescript-eslint-v8 brings in  ESLint v9 support and that was one of the blockers.

**Comment by MikeMcC399** [comment]
@bmish

Are you open to contributions to move the repo forward? It looks like you have been the major contributor all along.

As you know, `eslint-doc-generator` has been used (on your advice) to generate documentation for [eslint-plugin-cypress](https://github.com/cypress-io/eslint-plugin-cypress) and it's referred to in the [eslint-plugin-cypress/CONTRIBUTING.md](https://github.com/cypress-io/eslint-plugin-cypress/blob/master/CONTRIBUTING.md). So I'd like to help to ensure that `eslint-doc-generator` continues to be applicable by being reasonably up-to-date.

**Comment by bmish** [comment]
@MikeMcC399 yes I am open to contributions so please continue filing issues/PRs with any suggestions. I am busy traveling recently but would like to get things updated/modernized in this repository in the coming months, and any help will be appreciated.

**Comment by MikeMcC399** [comment]
@bmish 

Thanks very much for your positive reply! I'll make a start by highlighting some suggestions and I'll submit PRs as far as I am able to do.

**Comment by MikeMcC399** [comment]
@bmish 

- First PR in #530 for your review

**Comment by MikeMcC399** [comment]
- I submitted a set of dependency updates in Oct 2024, which have been merged and have moved the repo forward. I'm not able to tackle all open issues though (for example https://github.com/bmish/eslint-doc-generator/issues/545), so hopefully @bmish (or others) will be able to devote time to the open issues.

**Comment by bmish** [comment]
@MikeMcC399 really appreciate all the issues and PRs you filed.

I believe I have addressed everything related to updating the dev environment (Node, ESLint, TypeScript, and all other dependencies) for this repository. Please let me know if I have missed anything.

I do apologize for the delay, as there were a LOT of issues to untangle, and I was traveling and had other commitments the past few months.

This should enable much easier development going forward.

I'd like to release a major release soon to clear the pipeline with the breaking change of dropping all the old Node versions:
* https://github.com/bmish/eslint-doc-generator/pull/575

Are there any other breaking changes you had in mind that we should consider for inclusion with that? We don't necessarily need to include anything else, as nothing else is ready now, and we can perform additional major releases later when other things are ready. Just let me know what you plan to work on next or what you had in mind.

**Comment by MikeMcC399** [comment]
@bmish 

> really appreciate all the issues and PRs you filed.

I'm glad to have been able to assist!

> I believe I have addressed everything related to updating the dev environment (Node, ESLint, TypeScript, and all other dependencies) for this repository. Please let me know if I have missed anything.

You might want to reconsider

```json
  "peerDependencies": {
    "eslint": ">= 7"
  },
```

and bump that to `"eslint": ">= 8.57.1"` which is equivalent to `eslint@maintenance`. Users who need support for earlier (now unsupported) versions could just stay with the existing [eslint-doc-generator@1.7.1](https://github.com/bmish/eslint-doc-generator/releases/tag/v1.7.1) release.

> I do apologize for the delay, as there were a LOT of issues to untangle, and I was traveling and had other commitments the past few months.

I understand that the migration needed a lot of changes. I already came to that conclusion when I evaluated whether I could do it or not!
 
> This should enable much easier development going forward.
> 
> I'd like to release a major release soon to clear the pipeline with the breaking change of dropping all the old Node versions:
> 
> * [Require Node 18, 20, 22+ #575](https://github.com/bmish/eslint-doc-generator/pull/575)
> 
> Are there any other breaking changes you had in mind that we should consider for inclusion with that? We don't necessarily need to include anything else, as nothing else is ready now, and we can perform additional major releases later when other things are ready. Just let me know what you plan to work on next or what you had in mind.

I suggest you go ahead with the (breaking) release. I wasn't planning on any other contributions.

Good luck and thanks for the changes 🙂

**Comment by MikeMcC399** [comment]
- [eslint-doc-generator@2.0.0](https://github.com/bmish/eslint-doc-generator/releases/tag/v2.0.0) added `eslint@^9.0.0` to its package.json

https://github.com/bmish/eslint-doc-generator/blob/56c406f54ad059c32883e28f4a8d5cc27f00412e/package.json#L70

- Executing `npm install eslint@9 eslint-doc-generator@latest -g` generates no errors or deprecations

```text
├── eslint-doc-generator@2.0.2
├── eslint@9.18.0
```

This issue can now be closed, as it was focused on npm package environment compatibility and this has been resolved.

Kudos to @bmish for the hard work at the end of last year!
